### PR TITLE
Add isotope data to `MaterialParams`

### PR DESCRIPTION
**Pull request by stognini**
This PR covers one of the tasks defined in #801, which requires the addition of isotopic data into Celeritas. The main changes are:

- Expand `ImportData` to include isotope information and import it through the `GeantImporter`.
- Expand `celer-dump-data` accordingly.
- Add isotope information to `MaterialParamsData`.
- Expand `ElementView` to allow access to isotopic data, including a new `IsotopeView` class.
- Refactor `AtomicNumber` class to be used for both `A` and `Z` numbers.
- Expand Material test harness.

### CODE REVIEW

File: app/celer-dump-data.cc
Comment (sethrj) [inline app/celer-dump-data.cc]: Out of curiosity, is the `to_string` needed? It should be able to just reuse this "streamable" object.

File: app/celer-dump-data.cc
Comment (sethrj) [inline app/celer-dump-data.cc]: We don't bother printing the number fractions for elements in materials; should we do the same there? Or have a separate (more compact) table that has one row per iso/frac and "merges" the element name cell? Maybe for now just kill the relative abundances.

File: src/celeritas/Types.hh
Comment (sethrj) [inline src/celeritas/Types.hh]: I think the IDs in this section are sorted by name, and the "detailed" IDs should be as well (but it looks like DetectorId needs to be moved)

File: src/celeritas/ext/GeantImporter.cc
Comment (sethrj) [inline src/celeritas/ext/GeantImporter.cc]: Make a separate function `import_isotopes` that imports from `G4Isotope::GetIsotopeTable`. Otherwise you're potentially duplicating the same isotope.

File: src/celeritas/ext/GeantImporter.cc
Comment (sethrj) [inline src/celeritas/ext/GeantImporter.cc]: This definitely won't work if the same isotope is reused in multiple elements. Have you tried loading CMS with debug checking enabled?

File: src/celeritas/io/ImportElement.hh
Comment (sethrj) [inline src/celeritas/io/ImportElement.hh]: Should probably be `unsigned int` as in `ImportMatElemComponent` (maybe we should change that at some point?)

File: src/celeritas/mat/ElementView.hh
Comment (sethrj) [inline src/celeritas/mat/ElementView.hh]: The component ID is relative to this element, not the global array.
```suggestion
    CELER_EXPECT(id < this->num_isotopes());
```

File: src/celeritas/mat/ElementView.hh
Comment (sethrj) [inline src/celeritas/mat/ElementView.hh]: Same comment as above

File: src/celeritas/mat/MaterialParams.cc
Comment (sethrj) [inline src/celeritas/mat/MaterialParams.cc]: You shouldn't be `static_cast`ing an `int` to an `Id`; instead, cast to `unsigned int` and let the implicit conversion work. But if you take the recommendation above and make `element_id` an unsigned int, then you can just write 
```suggestion
                {elem_comp.element_id,
```

File: src/celeritas/mat/MaterialParams.cc
Comment (sethrj) [inline src/celeritas/mat/MaterialParams.cc]: Not sure why that was written as before... this should just be:
```suggestion
    return std::make_shared<MaterialParams>(std::move(input));
```

File: src/celeritas/mat/MaterialParams.cc
Comment (sethrj) [inline src/celeritas/mat/MaterialParams.cc]:
```suggestion
    isotope_data_ = !inp.materials.empty();
```
but can't you just replace `isotope_data_` in practice with `this->num_isotopes() > 0`?

File: src/celeritas/mat/MaterialParams.cc
Comment (sethrj) [inline src/celeritas/mat/MaterialParams.cc:144]: The extra conditional here is unnecessary: if no isotopes are provided, the loop will be empty and the labels will be a null-op.

File: src/celeritas/mat/MaterialParams.cc
Comment (sethrj) [inline src/celeritas/mat/MaterialParams.cc:172]: order the `ensure`s so isotopes are before elements (so that all the construction/data is always ordered from finest to coarsest)

File: src/celeritas/mat/MaterialParams.cc
Comment (sethrj) [inline src/celeritas/mat/MaterialParams.cc]: I don't think the conditional is needed here either. All of the operations will be null-op if it's empty.

File: src/celeritas/mat/MaterialParams.hh
Comment (sethrj) [inline src/celeritas/mat/MaterialParams.hh:88]: (Isotopes, elements, materials) or (materials, elements, isotopes)

File: src/celeritas/mat/MaterialParams.hh
Comment (sethrj) [inline src/celeritas/mat/MaterialParams.hh]:
```suggestion
    //! Whether isotopic data is present (may be false for EM-only physics)
    bool has_isotope_data() const { return this->num_isotopes() > 0; }
```
or maybe we should invert the bool and write it as "is_missing_isotopes"?

File: src/celeritas/mat/MaterialParams.cc
Comment (sethrj) [inline src/celeritas/mat/MaterialParams.cc]: This should be an assertion at the beginning:
```c++
CELER_EXPECT(inp.isotopes_fractions.empty() == this->has_isotope_data());
```

**Review by sethrj — CHANGES_REQUESTED** [review]

File: app/celer-dump-data.cc
Comment (stognini) [reply app/celer-dump-data.cc]: Fair. I'll remove the relative abundances. This will also resolve the previous question, about the `to_string` call (which was not really necessary, it was a remnant from a previous iteration that did not use stringstream).

File: src/celeritas/ext/GeantImporter.cc
Comment (stognini) [reply src/celeritas/ext/GeantImporter.cc]: Good point, I did not. However, the updated code with a separate `import_isotopes()` function works fine with `cms-hllhc-sd.gdml`, so we should be good.

File: src/celeritas/mat/MaterialParams.cc
Comment (stognini) [reply src/celeritas/mat/MaterialParams.cc]: Agreed, `num_isotopes()` provides the same functionality, so I'll remove the bool. I also realized that I accepted the wrong vscode autocompletion: that should have been `inp.isotopes`... 😬

**Comment by stognini** [comment]
After merging `develop`, I get a compilation warning

```shell
src/celeritas/ext/g4vg/SolidConverter.cc:98:6: warning: unused function 'calculate_theta_phi' [-Wunused-function]
auto calculate_theta_phi(G4ThreeVector const& axis) -> std::pair<double, double>
```

and 2 failed tests:

```shell
celeritas/ext/Vecgeom:SolidsTest.* (Failed)
celeritas/ext/Vecgeom:SolidsGeantTest.* (Failed)
```

**Comment by sethrj** [comment]
Ah shit, I was too quick on the merge button for #830. That's on me. I'll create a hotfix.

**Comment by sethrj** [comment]
Actually @stognini it's only the warning that's a problem. The failed tests are because your vecgeom version is too old: the failed test output will show the error and recommendation.

**Comment by stognini** [comment]
Yep, you're right. I spoke too soon. After looking at the test output and updating my vecgeom on spack the tests pass.

File: app/celer-dump-data.cc
Comment (sethrj) [inline app/celer-dump-data.cc:98]: I don't think this needs `to_string`. There's another `to_string(join(` in this file that I think is also unneeded.

File: src/celeritas/ext/GeantImporter.cc
Comment (sethrj) [inline src/celeritas/ext/GeantImporter.cc:260]: I think you may want to get the pointer first and then check if it's null. The way the isotope table works is if you `delete` an isotope, then the corresponding table entry is `nullptr`. So you should really be pre-sizing `isotopes`, getting the pointer, skipping the loop if it's null, and assigning to `ImportIsotope& isotope = isotopes[idx]`.

File: test/celeritas/mat/Material.test.cc
Comment (sethrj) [inline test/celeritas/mat/Material.test.cc]: These should have isotopes too. If you're providing any isotopes to MaterialParams you should provide all of them.

File: src/celeritas/mat/MaterialParams.cc
Comment (sethrj) [inline src/celeritas/mat/MaterialParams.cc]: I think there was some confusion about how this value is represented. I think you should just have `MevMass{nuclear_mass}` with no division by `c^2`. Check the value for carbon-12: `native_value_from(isotope.nuclear_mass) / celeritas::constants::atomic_mass` should be exactly 12.

**Review by sethrj — CHANGES_REQUESTED** [review]

File: test/celeritas/mat/Material.test.cc
Comment (stognini) [reply test/celeritas/mat/Material.test.cc]: Any specific reason to have isotopes in all of them other than consistency?
I've added them to a single element to test 2 scenarios at once:
1- Empty isotope cases
2- A small array of isotopes, to see if the MaterialParams construction and ElementView/IsotopeView are working as expected.
If both cases are satisfied, then any case (including real scenarios, with all elements having isotopes) should work.

File: app/celer-dump-data.cc
Comment (stognini) [reply app/celer-dump-data.cc:98]: I've added it to try to fix an odd formatting that I did not want to spend much time on. Without the `to_string(...)` the table looks like this:
```
| Element ID | Name | Atomic number | Mass (AMU) | Isotopes                                 |
| ---------- | ---- | ------------- | ---------- | ---------------------------------------- |
| 0          | O    | 8             | 16         | O16                                     , O17, O18 |
| 1          | Pb   | 82            | 207        | Pb204                                   , Pb206, Pb207, Pb208 |
```

File: src/celeritas/mat/MaterialParams.cc
Comment (stognini) [reply src/celeritas/mat/MaterialParams.cc]: Ohh I misread the `MevMass = Quantity<MevPerCsq>;`. The unit _is_ `MeV/c^2` and thats equivalent to `MeV` in G4. Not that we are storing the value after dividing per c^2. I'll fix that.

> `native_value_from(isotope.nuclear_mass) / celeritas::constants::atomic_mass` should be exactly 12

Yes! Agreed.

Now, this point made me double check the `G4NucleiProperties::GetNuclearMass(...)`, which has 2 overloads:
```cpp
GetNuclearMass(G4double A, G4double Z);
GetNuclearMass(G4int A, G4int Z);
```
Which are slightly different. The overload with doubles is more precise, so I should use that one. @whokion do you also agree with the need to use the one with doubles as inputs?

EDIT: interesting, `G4Isotope` only provides `Z` and `A` as `G4int`, whereas `G4Element` returns both as `G4double`... so I guess we're limited to using the second overload with `G4ints` as inputs for isotopes...

Comment (whokion) [reply src/celeritas/mat/MaterialParams.cc]: Yes, @stognini, you  are right and should use `GetNuclearMass(G4double A, G4double Z)` as it also covers `GetNuclearMass(G4int A, G4int Z)`, (i.e., using the known PDF mass or table values when A is close to an integer, otherwise, taking into account an estimated mass access, binding energy, etc).

Comment (sethrj) [reply src/celeritas/mat/MaterialParams.cc]: @stognini Looking at the source code, I think you want to use the `int` function signature because that uses a table lookup for the data rather than a rough functional approximation.

Comment (stognini) [reply src/celeritas/mat/MaterialParams.cc]: The current implementation is using the `G4int` one (since isotopes do not provide `Z` and `A` as doubles). But to me it seems that the double is more precise, as it calls the `int` function signature if `fabs((double)A - (int)A) < 1e-10`, and thus uses table lookup, otherwise it uses the calculated approximation. Nevertheless, unless new accessors are added to `G4Isotope`, we don't really have a choice here...

Comment (sethrj) [reply src/celeritas/mat/MaterialParams.cc]: Z and A shouldn't ever be doubles. It looks like the `int, int` method falls back to the functional approximation if it's not in any of the tables.

The whole "z as a double" reminds me of `if (Z < 1.5)`... 😠

Comment (stognini) [reply src/celeritas/mat/MaterialParams.cc]: Oh man, you're right. I missed the `NuclearMass(...)` at the bottom when skimming through it...

Yeah, G4 accessors are weird. For `A` you only get a double that represents the effective number of nucleons, which is pretty much the same as giving you its amu, so that's why it returns a double. For `Z` it should be `int` for sure.

**Review by hhollenb — APPROVED** [review]
For implementing an IsotopeSelector, would we care about having information about the isotope stability as well, or just sample from abundances?

Regardless this looks good to me.

File: app/celer-dump-data.cc
Comment (sethrj) [reply app/celer-dump-data.cc:98]: Ah, OK. That makes sense: it has to do with `setw` being passed to "the next streamed element". std streams are not the best design...

File: test/celeritas/mat/Material.test.cc
Comment (sethrj) [reply test/celeritas/mat/Material.test.cc]: The reason is to add the assertion
```
CELER_EXPECT(inp.isotopes_fractions.empty() == this->has_isotope_data());
```
since I think we only want the cases of "all isotope data is available" and "no isotope data is available". You don't want your isotope sampler to work for some elements but not others. The only time we want to support "no isotope data" is if we're using only simple EM physics.

**Review by sethrj — APPROVED** [review]
This looks perfect! Since it's still compatible with the existing ROOT files we can definitely include this in 0.3.1.

**Comment by stognini** [comment]
> would we care about having information about the isotope stability as well, or just sample from abundances?

I don't remember seeing any HEP MC simulation where people included the detector composition having decaying isotopes generating any background signal. Not even sure if Geant4 has any mechanism to deal with that, since as far as I know element information is constant and isotope decay means that the fractional abundances are changing over time (and new elements are probably being added to the initial list).

File: src/celeritas/mat/MaterialParams.cc
Comment (whokion) [reply src/celeritas/mat/MaterialParams.cc]: Okay, I was wrong and agree that we 'd better use `GetNuclearMass(G4int A, G4int Z)` for storing the nuclear mass for the isotope case (as the result is same anyway even if GetNuclearMass(G4double A, G4double Z) is used with int A and int Z.  The user case for GetNuclearMass(G4double A, G4double Z) may be only for using the effective (or weighted) atomic mass number of an element, which is irrelevant for this importing.  Also wondering what real use cases are for using GetNuclearMass(G4double A, G4double Z) in any application, in which may need to use this on-the-fly calculation.

**Comment by whokion** [comment]
> > would we care about having information about the isotope stability as well, or just sample from abundances?
> 
> I don't remember seeing any HEP MC simulation where people included the detector composition having decaying isotopes generating any background signal. Not even sure if Geant4 has any mechanism to deal with that, since as far as I know element information is constant and isotope decay means that the fractional abundances are changing over time (and new elements are probably being added to the initial list).

Probably in space or medical applications.  Anyway, it is still a good question/point as there exist unstable isotopes even though their life times may be much long than usual HEP applications.

**Comment by sethrj** [comment]
@stognini I realized while looking at the "todos" in our material params that the plan was to name these *nuclides* which is more correct nomenclature than *isotopes* (see https://github.com/celeritas-project/celeritas/issues/801#issuecomment-1583150560 also). Isotopes are only defined with respect to a given element, but the data we have is for a *nuclide* which includes Z data and unique properties about the binding energy etc. Do you recall whether we decided to adopt the geant4 nomenclature deliberately, or whether it just accidentally snuck in?

**Comment by stognini** [comment]
Ohh I had forgotten about this discussion. I don't remember what was the conclusion, but I agree that nuclide is more general and makes more sense. I can refactor the needed classes to use nuclide.

**Comment by sethrj** [comment]
@stognini Don't worry about it yet; I've added a note and we can put it on the nomenclature to-do for v1.0.

**Pull request merged**
